homeassistant/components/mqtt/light/schema_basic.py

```diff
@@ -817,7 +818,9 @@ def set_optimistic(
             # Make sure the brightness is not rounded down to 0
             device_brightness = max(device_brightness, 1)
             command_tpl = self._command_templates[CONF_BRIGHTNESS_COMMAND_TEMPLATE]
-            device_brightness_payload = command_tpl(device_brightness, None)
+            device_brightness_payload = command_tpl(
+                device_brightness, {"brightness_provided": brightness_provided}
+            )
```

Comment:
May be we can just pass `kwargs` as parameter array, and we could do for every command template? @emontnemery what do you think?

Reply:
@jbouwh That is maybe also a possiblity, but in that case it is needed to make a copy of the complete `kwargs` dict _before_ making any modifications, like done at: https://github.com/home-assistant/core/blob/928342c7d5dcb7406928667de2d83ee47534ce90/homeassistant/components/mqtt/light/schema_basic.py#L759 

Otherwise the information that is currently provided is missing. Since to goal is to know if the brightness was provided by the start of the action or not. 

Other impact is that the documentation needs to be updated with all available arguments _and_ this needs to be kept in sync if in the future more arguments will be added. So maybe this makes it more complex for a documentation point of view.

Reply:
If `kwargs` is used a source that `value` will still be available. In the template you can check for the existance of any parameter:
`https://www.home-assistant.io/docs/configuration/templating/#is-defined. But in this case `brightness` will be set always as there is a condition check.

Reply:
I'm familiar with Jinja2 templating, but thanks for the information.

Problem is indeed that the `kwargs` object is changed depending on the other options (the line I refer to) and later on used for other conditions (the lines you refer to). So `brightness` is always set in this situation and therefore not usable.

So only when a copy of the dict is made at the beginning or the dict is not changed anymore it will work.

Reply:
`kwargs` is not changed and could be used as parameter array for to command templates IMO.

Reply:
The problem occurs when `on_command_type` is configured to `brightness` and a `on` action is done (so no brightness provided). In that case this condition is true: https://github.com/home-assistant/core/blob/928342c7d5dcb7406928667de2d83ee47534ce90/homeassistant/components/mqtt/light/schema_basic.py#L751-L759 and the `kwargs` object is changed, because the `brightness` key is added. This is done before the template is processed.

To prevent I'm wrong, I will do a small test later today. Thanks for your quick replies.

Reply:
Okay that makes sense, lets see what @emontnemery thinks

Reply:
It's not clear to me why this is needed, it seems the light has a problem because when you attempt to turn it on, the brightness is set to max instead of to the previous brightness?

Have you raised an issue on https://github.com/zwave-js/zwave-js-ui about why a light can't be turned back on to it's previous brightness?

In the test case, you send 255 to the light's brightness topic, is that a correct example for resuming a `zwave-js-ui` light's previous brightness?

Reply:
@emontnemery I starten with a question on the Z-Wave JS UI project, since I my first assumption was an incorrect command type was used https://github.com/zwave-js/zwave-js-ui/discussions/3030. But that was something that was tested in the past and results in other isused.

The interesting part was that it work as expected with Domoticz. So therefore I analyzed the differences and proposed this easy and non breaking solution, see here the details: https://community.home-assistant.io/t/mqtt-light-brightness-to-previous-value-when-using-the-on-button/557645.

Makes this the reason clear?

Regarding the question of @jbouwh about making even more parameters available. After thinking a bit longer I suggest to keep the interface simple and only add when something is needed. This because we see that also other projects, like Domoticz, implement the Home Assistant autodiscovery protocol. Adding a parameter is just a minor feature, removing or renaming something in the future is a breaking change. What is your opinion?

Reply:
No, it's unfortunately very unclear still.

You gave this configuration as an example:

```yaml
command_topic: zwave/95/38/1/targetValue/set
state_topic: zwave/95/38/1/currentValue
state_value_template: '{{ "OFF" if value_json.value == 0 else "ON" }}'
brightness_command_topic: zwave/95/38/1/targetValue/set
brightness_scale: 99
brightness_state_topic: zwave/95/38/1/currentValue
brightness_value_template: '{{ value_json.value }}'
on_command_type: brightness
device:
  ...
name: Spots bureau werkkamer_dimmer_1
unique_id: zwavejs2mqtt_0xf1858849_95-38-1-currentValue
platform: mqtt
```

I believe the reason this does not work is that we get a state update from the light when it's turned off, where it says it's brightness is 0.

@jbouwh A brightness of 0 is not a valid brightness in Home Assistant. Should we perhaps trap that in `brightness_received`?

Reply:
No problem, it is a lot of information, so lets explain it shortly and hopefully simple :)

In the mentioned example the `on_command_type` is configured to `brightness`, which means that only a single command will be sent, instead of a on/off command AND a seperated brightness command. 
The configuration created by Z-Wave JS UI set the `brightness_command_topic` and the `command_topic` to the same value.

When a `on` action is executed in Home Assistant the `async_turn_on` function is called _without_ a `brightness` value. In that case the following function https://github.com/home-assistant/core/pull/91125/files#diff-5349a6ce9c2c784e9fd69377eec96061257d838e33c937e671e9f99aa5ddba06R751-R759 set the brightness to 255. So this results in sending 99 (because of the `brightness_scale`) to the `brightness_command_topic`

When a `on` action is executed in Domoticz also no brightness is available, but Domoticz in sending in that case only the `payload_on` to the `command_topic` . This is the behaviour I expect, because the ON message will restore the last brightness.

In case of an `off` action only the payload off message will be sent to the `command_topic` . This is the same for Home Assistant and Domoticz.

I was looking for a solution to send payload_on ("ON") or 255 in case of an on action in a not breaking way. But now I have the feeling that the implementation in Home Assistant for the "ON" action is not correct and we should fix the issue by just calling the `command_topic` with payload_on in case no brightness is available? So actually the same a Domoticz does. 

Is it clear now and what is your opinion?

Reply:
@emontnemery You are right that a brightness value of 0 is received when the light is switched off. I will try your suggestion to see if Home Assitant in that case remembers the correct brightness value, in that case that is also a solution.

Reply:
@emontnemery I tested your solution by changing the configuration and adapting the template:
```
brightness_value_template: '{{ value_json.value if value_json.value > 0 else none }}'
```

This solves my problem in Home Assistant and the brightness is now correctly set to the previous version! There are two points to mention:
- By using this template Domoticz is not working anymore, as they are not correctly parsing the templates (this is not a problem of Home Assistant ofcourse).
- The first time, when Home Assistant does not know the previous state, it will be set to full brightness. But this is expected behaviour (and can be solved by using the retain flag in MQTT).

Conclusion, the implementation I did in this pull request is not the right direction. I hope @jbouwh agrees and we can see the brightness of 0 as in invalid value in the brightness_received function. That is a solution that will not require any changes on the Z-Wave JS UI project (so will also not break anything for Domoticz).

What will be the next steps (as this is my first pull request for this project): do we need to decline this pull request or can I update this pull request with the suggested solution after agreement of @jbouwh?

Thank you very much for helping me with this issue.

Reply:
This https://github.com/home-assistant/core/pull/91296 should avoid a zero brightness to be processed (but will still log a warning).

Reply:
> By using this template Domoticz is not working anymore, as they are not correctly parsing the templates (this is not a problem of Home Assistant ofcourse).

That's very much not a problem in Home Assistant, and in my opinion Z-Wave JS UI should use a template like that.
Does Domoticz have built in support for MQTT discovery these days, or are you using a plugin for that purpose?

Reply:
I totally agree and I think work needs to be done in the Domoticz part. The MQTT Discovery is build in and new way of working for Z-Wave devices: https://www.domoticz.com/wiki/Zwave-JS-UI

The Jinjia 2 templating support is not implemented completely the are manually processing the templates which results in very limited options. See https://github.com/domoticz/domoticz/blob/2023.1/hardware/MQTTAutoDiscover.cpp

So I totally agree the problem is Domoticz, but do you see any issue with ignoring the zero brightness value in Home Assitant as already implemented by @jbouwh? Or do you seen any usecase for accepting value 0?

I'm searching for an easy solution that wil not directly breaks Z-Wave JS UI with Domoticz. Because the now I know that when I propose the Z-Wave JS UI project to update the template all Domoticz users have an issue.